PSP: improve parsing and validation of SELinux levels

php-coder opened the pull request:
**What this PR does / why we need it**:
At this moment, when we're comparing SELinux levels of PSP and container, we compare them as strings. We don't take into account that categories in a level may be specified in a different order (for example, `s0:c0,c6` is the same as `s0:c6,c0`).

This PR improves handling of SELinux levels by doing logical comparison.

**Special notes for your reviewer**:
Here is the issue in OpenShift tracker from @pweil-: https://github.com/openshift/origin/issues/15627
Relate PR to fixing this in OpenShift: https://github.com/openshift/origin/pull/16432

**Release note**:
```release-note
NONE
```

PTAL @pweil- 
CC @simo5

xiangpengzhao commented:
/ok-to-test

php-coder commented:
/hold

php-coder commented:
/unhold

php-coder commented:
/hold cancel

php-coder commented:
@smarterclayton @brendandburns PTAL.

php-coder commented:
/test pull-kubernetes-unit

brendandburns commented on pkg/security/podsecuritypolicy/selinux/mustrunas.go:
```diff
@@ -74,3 +81,128 @@ func (s *mustRunAs) Validate(fldPath *field.Path, _ *api.Pod, _ *api.Container,
+		return false, err
+	}
+
+	if !reflect.DeepEqual(expectedSensitivity, actualSensitivity) {
```
Do we really want to use DeepEqual here, this uses reflection.

You can just use `==` for arrays in golang.

See:
https://golang.org/ref/spec#Comparison_operators

"Array values are comparable if values of the array element type are comparable. Two array values are equal if their corresponding elements are equal."

php-coder replied:
Unfortunately, we can't use `==` here because we're comparing slices and not arrays. These slices are came from invocation of `strings.Split()` and may be differ in length.

smarterclayton replied:
I would probably prefer a custom equality function over reflect.DeepEqual.  You are checking that two string arrays are equal, better to have an explicit method `stringSliceEqual` or something

php-coder replied:
Added in the last update.

brendandburns commented on pkg/security/podsecuritypolicy/selinux/mustrunas.go:
```diff
@@ -74,3 +81,128 @@ func (s *mustRunAs) Validate(fldPath *field.Path, _ *api.Pod, _ *api.Container,
+
+	// TODO: add support for dominance
+	// See: https://selinuxproject.org/page/NB_MLS#Managing_Security_Levels_via_Dominance_Rules
+	return reflect.DeepEqual(expectedCategories, actualCategories), nil
```
here too wrt array comparison.

brendandburns commented on pkg/security/podsecuritypolicy/selinux/mustrunas.go:
```diff
@@ -74,3 +81,128 @@ func (s *mustRunAs) Validate(fldPath *field.Path, _ *api.Pod, _ *api.Container,
+
+func canonicalizeItems(str, itemsSeparator, rangeSeparator, boundaryPrefix string) ([]string, error) {
+	parts := strings.Split(str, itemsSeparator)
+	result := make([]string, 0, len(parts))
```
nit: I personally wouldn't bother pre-allocating here, but your call.

php-coder replied:
I'd prefer to leave it.

brendandburns commented on pkg/security/podsecuritypolicy/selinux/mustrunas.go:
```diff
@@ -74,3 +81,128 @@ func (s *mustRunAs) Validate(fldPath *field.Path, _ *api.Pod, _ *api.Container,
+	}
+
+	// although sorting digits as strings is leading to a wrong order,
+	// it doesn't matter because we only need to sort both parts in a similar way
```
replace 'similar' with 'identical' for precision in language.

(alternately 'consistent')

php-coder replied:
Fixed.

brendandburns commented on pkg/security/podsecuritypolicy/selinux/mustrunas.go:
```diff
@@ -74,3 +81,128 @@ func (s *mustRunAs) Validate(fldPath *field.Path, _ *api.Pod, _ *api.Container,
+
+	// Level format: https://selinuxproject.org/page/MLSStatements#level
+
+	// "s0:c6,c0" => [ "s0", "c6,c0" ]
```
All this parsing seems super brittle to me, and I'd really prefer that it was done with a regular expression...



Not going to require it, but maybe at least a TODO?

php-coder replied:
Ok, I've added a TODO.

php-coder replied:
In future it should be possible to use `CanonicalizeContext()` from selinux: https://github.com/opencontainers/selinux/pull/19 Here is also a related discussion: https://github.com/openshift/origin/pull/16432#issuecomment-338716218

php-coder commented:
PR was updated to address the comments. No code changes were made (only code comments).

@brendandburns PTAL

smarterclayton commented on pkg/security/podsecuritypolicy/selinux/mustrunas.go:
```diff
@@ -74,3 +81,129 @@ func (s *mustRunAs) Validate(fldPath *field.Path, _ *api.Pod, _ *api.Container,
 	return allErrs
 }
+
+func equalLevels(expectedLevel, actualLevel string) (bool, error) {
```
godoc for this

smarterclayton commented on pkg/security/podsecuritypolicy/selinux/mustrunas.go:
```diff
@@ -74,3 +81,129 @@ func (s *mustRunAs) Validate(fldPath *field.Path, _ *api.Pod, _ *api.Container,
+	return begin, end, nil
+}
+
+// Parses a string with a range boundary in a format "$prefix$value" by extracting value and converting it to an integer.
```
proper godoc style

smarterclayton commented on pkg/security/podsecuritypolicy/selinux/mustrunas.go:
```diff
@@ -74,3 +81,129 @@ func (s *mustRunAs) Validate(fldPath *field.Path, _ *api.Pod, _ *api.Container,
+	parts := strings.Split(str, itemsSeparator)
+	result := make([]string, 0, len(parts))
+
+	for _, value := range parts {
```
Do we *really* need to support range?  It's fairly complex, and seems unlikely that it will make much difference to users.

smarterclayton replied:
Specifically do we really need to support range expansion

php-coder replied:
@smarterclayton No one is answered and I don't know the answer too. So what we will we do with this PR?

smarterclayton replied:
@eparis @pweil- can we think of a concrete reason why we *need* range expansion in Kube?

php-coder replied:
@smarterclayton Would this PR be approved if I remove range support?

Here is what @pweil- [said](https://github.com/openshift/origin/issues/15627#issuecomment-330307568):
> Having a range check would be nice to have but I think it's less likely to cause an error in the real world.

pweil- replied:
I'm in favor or removing it if it is holding this up.  It can be a follow up if we find folks need it.

eparis replied:
I don't know where the inputs come from here. The kernel will always use ranged values if possible. If we randomly pick `c0,c1,c2` the kernel will always report that as `c0..c2`. If we ever expect to do comparison with any value that did a round-trip through the kernel then range comparison is required :-(

php-coder commented:
@brendandburns @smarterclayton Comments were addressed (except the one where I don't know the answer). PTAL.

brendandburns commented:
LGTM for me, but I'll wait to see if Clayton has comments before stamping.

php-coder commented:
PTAL @smarterclayton

php-coder commented:
- pull-kubernetes-e2e-gce-device-plugin-gpu failed because of https://github.com/kubernetes/kubernetes/issues/56737 or https://github.com/kubernetes/kubernetes/issues/56739
- pull-kubernetes-e2e-gce failed because of https://github.com/kubernetes/kubernetes/issues/56913

php-coder commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-e2e-gce

@smarterclayton PTAL

php-coder commented:
@smarterclayton Please take a look.

brendandburns commented:
/retest

php-coder commented:
`pull-kubernetes-e2e-gce-device-plugin-gpu` fails with

> I0103 01:21:13.501] Jan  3 01:01:23.240: INFO: Error while reading data from e2e-11702-minion-group-4991: Get https://104.198.9.160/api/v1/nodes/e2e-11702-minion-group-4991:10250/proxy/stats/summary: context deadline exceeded

that most likely caused by

> I0103 01:21:14.190] Jan  3 01:06:08.487: INFO: At 2018-01-03 00:56:08 +0000 UTC - event for nvidia-driver-installer: {daemonset-controller } FailedPlacement: failed to place pod on "e2e-11702-master": **Node didn't have enough resource**: cpu, requested: 150, used: 982, capacity: 1000

php-coder commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

php-coder commented:
Test `[Feature:GPUDevicePlugin] run Nvidia GPU Device Plugin tests on Container Optimized OS only` is failing to always:
```
test/e2e/scheduling/nvidia-gpus.go:230
Timed out after 600.000s.
Expected
    <bool>: false
to be true
test/e2e/scheduling/nvidia-gpus.go:200
```
@kubernetes/sig-scheduling-test-failures Could you look and advice/help here? Thanks!

php-coder commented:
@smarterclayton PTAL

/test pull-kubernetes-e2e-gce-device-plugin-gpu

php-coder commented:
@smarterclayton It's green. Please, take a look or advise another approver.

Commit:
PSP: when comparing categories in SELinux levels, ignore its order.

Commit:
Update autogenerated files.

php-coder commented:
I rewrote the PR to make it simple, short and fix only one thing: the original issue when a pod was rejected if SELinux levels don't match:

```console
$ kubectl create -f selinux-pod.yaml 
Error from server (Forbidden): error when creating "selinux-pod.yaml": pods "testing-selinux-" is forbidden: unable to validate against any pod security policy: [spec.containers[0].securityContext.seLinuxOptions.level: Invalid value: "s0:c6,c0": must be s0:c0,c6]
```

@pweil- @smarterclayton @brendanburns PTAL

simo5 commented:
/lgtm

pweil- reviewed (commented):
@php-coder the initial validate tests were pretty coarse grained since we were doing a simple comparison.  Do you think it's worth expanding that coverage now even though the message is the same?

I would also have some tests for your utility method, I'm really surprised that doesn't exist somewhere already.

Other than that this is LGTM

Commit:
selinux/mustrunas_test.go(TestMustRunAsValidate): rename a member to make its meaning obvious.

Commit:
selinux/mustrunas_test.go(TestMustRunAsValidate): make PSP SeLinux options configurable.

Commit:
selinux/mustrunas_test.go(TestMustRunAsValidate): add more test cases to improve code coverage.

php-coder commented:
@pweil- Thanks for spotting this! I updated PR and added more tests to improve code coverage.

As for the message, I don't see a value to change it because we're adding expected level to the error message in the caller:

> [spec.containers[0].securityContext.seLinuxOptions.level: Invalid value: "s0:c6,c0": must be s0:c0,c6]

php-coder commented:
@tallclair PTAL

pweil- reviewed (approved):
/lgtm

Thanks for adding the tests

tallclair commented on pkg/security/podsecuritypolicy/util/util.go:
```diff
@@ -222,3 +222,17 @@ func hasPathPrefix(s, pathPrefix string) bool {
+
+// EqualStringSlices compares string slices for equality. Slices are equal when
+// their sizes and elements on similar positions are equal.
+func EqualStringSlices(a, b []string) bool {
```
Is this faster than `reflect.DeepEqual`? Otherwise, just use DeepEqual.

php-coder replied:
I believe that yes, it's faster. Also this is what @brendanburns and @smarterclayton asked me to do earlier: https://github.com/kubernetes/kubernetes/pull/52799#discussion_r146095101

tallclair commented:
/lgtm

tallclair commented:
/approve no-issue

(Actual issue: https://github.com/openshift/origin/issues/15627)
